### PR TITLE
netbeans-479: Added ConvertToVarHint to replace explicit type with var

### PR DESCRIPTION
**Pull request by arusinha**
The IDE will include a prompt to replace the explicit types with ‘var’, when the type can be inferred from the context.

e.g..,  String str = new String("Hello");   

can be replaced with  var str = new String("hello");

Preconditions for the hint:
1. hint will be enabled only for JDK10 or higher
2. hint willwork only inside local scope
3. hint will be disabled if var type is already present.
4.  hint will not support generic array.
  e.g.,  Arraylist<String>[] arr = new ArrayList<String>[5];

Apache Jira Link:
https://issues.apache.org/jira/browse/NETBEANS-479

Explicit type of local variable will be converted to var type if  type of variable is same as assignment type
Below Scenarios are covered:
1. Object Initialization:
    Integer i = new Integer("10");
    Fix : var i = new Integer("10");
	
   HashMap<Integer,Integer> map= new HashMap<Integer,Integer>();
   Fix : var map= new HashMap<Integer,Integer>();   

2.  Array Initialization
   int[][]arr = new int[3][]; 
   Fix :  var arr = new int[3][];
   
3. Assignment of literal:
   int a = 10;
   Fix:    var a = 10;
   
4.  Assignment to a method call
 var list = Collections.unmodifiableList(new ArrayList<String>()) ;

Below Assignment scenarios are not covered.
1. Anonymous Object  
2. Lambda expression
3. Diamond interface

### CODE REVIEW

File: java.hints/src/org/netbeans/modules/java/hints/suggestions/ConvertToVarHint.java [hunk 1]
```diff
@@ -0,0 +1,449 @@
...
+    @Messages("MSG_ConvertibleToVarType=Explict type can be replaced with 'var'")  //NOI18N
+    public static ErrorDescription checkNewObjInit(HintContext ctx) {
+
+        TreePath treePath = ctx.getVariables().get("$var");
+
+        if (treePath == null) {
+            return null;
+        }
+
+        if (!isHintEnabled(ctx)) {
```
Comment (emilianbold) [inline java.hints/src/org/netbeans/modules/java/hints/suggestions/ConvertToVarHint.java]: Shouldn't isHintEnabled be the first check?

File: java.hints/src/org/netbeans/modules/java/hints/suggestions/ConvertToVarHint.java [hunk 2]
```diff
@@ -0,0 +1,449 @@
...
+        }
+
+        Tree l = initPath.getLeaf();
+
+        // checks rhs is a literal or not
+        if (!(isLiteral(l))) {
+            return null;
+        }
+
+        if (!isHintEnabled(ctx)) {
```
Comment (emilianbold) [inline java.hints/src/org/netbeans/modules/java/hints/suggestions/ConvertToVarHint.java]: Same here why is isHintEnabled checked so late?

File: java.hints/src/org/netbeans/modules/java/hints/suggestions/ConvertToVarHint.java [hunk 3]
```diff
@@ -0,0 +1,449 @@
...
+     *
+     * @param ctx : HintContext
+     * @return ErrorDescription if variable initializer is Lambda expression.
+     */
+    @Hint(displayName = "#DN_CanUseVarForLambdaExpRef", description = "#DESC_CanUseVarForLambdaExpRef", category = "suggestions")
+    @TriggerTreeKind(Kind.LAMBDA_EXPRESSION)
+    public static ErrorDescription checkLambdaExpr(HintContext ctx) {
+        TypeMirror samType = ctx.getInfo().getTrees().getTypeMirror(ctx.getPath());
+
+        if (!isHintEnabled(ctx)) {
```
Comment (emilianbold) [inline java.hints/src/org/netbeans/modules/java/hints/suggestions/ConvertToVarHint.java]: ... and here.

File: java.hints/src/org/netbeans/modules/java/hints/suggestions/ConvertToVarHint.java [hunk 4]
```diff
@@ -0,0 +1,449 @@
...
+     * @return ErrorDescription if variable initializer Object uses diamond
+     * operator(<>).
+     */
+    @Hint(displayName = "#DN_CanUseVarForDiamondInterfaceRef", description = "#DESC_CanUseVarForDiamondInterfaceRef", category = "suggestions")
+    @TriggerPattern("new $type<$T$>($params$)") //NOI18N
+    public static ErrorDescription checkdiamondType(HintContext ctx) {
+
+        TreePath treePath = ctx.getPath();
+
+        if (!isHintEnabled(ctx)) {
```
Comment (emilianbold) [inline java.hints/src/org/netbeans/modules/java/hints/suggestions/ConvertToVarHint.java]: and here and the next one.

File: java.hints/test/unit/src/org/netbeans/modules/java/hints/suggestions/ConvertToVarHintTest.java [hunk 5]
```diff
@@ -0,0 +1,346 @@
...
+package org.netbeans.modules.java.hints.suggestions;
+
+import org.junit.Test;
+import org.netbeans.modules.java.hints.test.api.HintTest;
+
+/**
+ *
+ * @author arusinha
+ */
+public class ConvertToVarHintTest {
```
Comment (emilianbold) [inline java.hints/test/unit/src/org/netbeans/modules/java/hints/suggestions/ConvertToVarHintTest.java]: Big +1 for the tests!

Comment (arusinha) [reply java.hints/test/unit/src/org/netbeans/modules/java/hints/suggestions/ConvertToVarHintTest.java]: This is addressed in commit https://github.com/apache/incubator-netbeans/pull/463/commits/b733846a4ab3081e2b4d56531ab3c3d041fdd93e

File: java.hints/src/org/netbeans/modules/java/hints/suggestions/ConvertToVarHint.java (hunk 2 above)
Comment (arusinha) [reply java.hints/src/org/netbeans/modules/java/hints/suggestions/ConvertToVarHint.java]: https://github.com/apache/incubator-netbeans/pull/463/commits/b733846a4ab3081e2b4d56531ab3c3d041fdd93e

File: java.hints/src/org/netbeans/modules/java/hints/suggestions/ConvertToVarHint.java (hunk 1 above)
Comment (arusinha) [reply java.hints/src/org/netbeans/modules/java/hints/suggestions/ConvertToVarHint.java]: this is addressed in commit : https://github.com/apache/incubator-netbeans/pull/463/commits/b733846a4ab3081e2b4d56531ab3c3d041fdd93e

File: java.hints/src/org/netbeans/modules/java/hints/suggestions/ConvertToVarHint.java (hunk 3 above)
Comment (arusinha) [reply java.hints/src/org/netbeans/modules/java/hints/suggestions/ConvertToVarHint.java]: this is addressed in commit : https://github.com/apache/incubator-netbeans/pull/463/commits/b733846a4ab3081e2b4d56531ab3c3d041fdd93e

File: java.hints/src/org/netbeans/modules/java/hints/suggestions/ConvertToVarHint.java (hunk 4 above)
Comment (arusinha) [reply java.hints/src/org/netbeans/modules/java/hints/suggestions/ConvertToVarHint.java]: this is addressed in commit : https://github.com/apache/incubator-netbeans/pull/463/commits/b733846a4ab3081e2b4d56531ab3c3d041fdd93e

**Comment by jlahoda** [comment]
Hi,

Sorry for delay. Some initial comments:
-please note that every occurrence of @Hint means a separate "hint" and a separate entry in Tools/Options/Editor/Hints. I think there should only be a single @Hint for this (on the class), which means a single hint/single configuration entry, and then there can be any number of @Trigger methods inside the class that will all work for that single hint. Some conversion may be seen undesirable (worsening readability), so it might be desirable to allow disabling the conversion in some contexts, although I am not sure offhand for which it should be. Defining simple boolean options should be simple, see @BooleanOption, e.g. in: java.hints/src/org/netbeans/modules/java/hints/perf/SizeEqualsZero.java
-please note that @Hint has minSourceVersion that can be used to specify minimal -source version for the hint. But even if that didn't work in a particular case, there is CompilationInfo.getSourceVersion(), which should be used instead of SourceLevelQuery in hints.
-category = "suggestions" -> category = "rules15" (the category name is confusing, but it is not easy to change; this will be under "JDK Migration Support")
-I wonder if the hint shouldn't have simply @TriggerTreeKind(VARIABLE) or @TriggerPattern("$mods$ $type $var =new $init;") and then internally trying to figure out if the variable can be var or not. (Also please note there will soon be an API method to check if a variable is "var" or not that the hint will be able to use.)

**Comment by jlahoda** [comment]
Ok, I think. We can refine later if needed. Please note the branch cannot be merged - it will be necessary to rebase to/merge with master and manually resolve any conflicts.

**Comment by geertjanw** [comment]
OK, thanks, merging.

**Pull request merged**